node/pkg/watchers/ibc/watcher.go

```diff
@@ -0,0 +1,734 @@
+			Help: "Current slot height on an IBC connected chain (the block height on wormchain)",
+		}, []string{"chain_name"})
+	invalidChainIdMismatches = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "wormhole_ibc_chain_id_mismatches",
+			Help: "Total number of cases where the wormhole chain ID does not match the IBC connection ID",
+		}, []string{"ibc_channel_id"})
+)
+
+type (
+	// Watcher is responsible for monitoring the IBC contract on wormchain and publishing wormhole messages for all chains connected via IBC.
+	Watcher struct {
+		wsUrl           string
+		lcdUrl          string
+		contractAddress string
+		logger          *zap.Logger
+
+		// chainConfig is the list of chains to be monitored, along with their channel data.
+		chainConfig ChainConfig
+
+		// channelMap provides access by IBC connection ID.
+		connectionMap map[string]*connectionEntry
+
+		// chainMap provides access by chain ID.
+		chainMap map[vaa.ChainID]*connectionEntry
+
+		// connectionIdMap provides a mapping from IBC channel ID to IBC connection ID.
+		connectionIdMap map[string]string
+
+		// connectionIdLock protects connectionIdMap.
+		connectionIdLock sync.Mutex
+	}
+
+	// connectionEntry defines the chain that is associated with an IBC connection.
+	connectionEntry struct {
+		connectionID string
+		chainID      vaa.ChainID
+		chainName    string
+		readiness    readiness.Component
+		msgC         chan<- *common.MessagePublication
+		obsvReqC     <-chan *gossipv1.ObservationRequest
+	}
+)
+
+// NewWatcher creates a new IBC contract watcher
+func NewWatcher(
+	wsUrl string,
+	lcdUrl string,
+	contractAddress string,
+	chainConfig ChainConfig,
+) *Watcher {
+	return &Watcher{
+		wsUrl:           wsUrl,
+		lcdUrl:          lcdUrl,
+		contractAddress: contractAddress,
+		chainConfig:     chainConfig,
+		connectionIdMap: make(map[string]string),
+		connectionMap:   make(map[string]*connectionEntry),
+		chainMap:        make(map[vaa.ChainID]*connectionEntry),
+	}
+}
+
+// clientRequest is used to subscribe for events from the contract.
+type clientRequest struct {
+	JSONRPC string `json:"jsonrpc"`
+	// A String containing the name of the method to be invoked.
+	Method string `json:"method"`
+	// Object to pass as request parameter to the method.
+	Params [1]string `json:"params"`
+	// The request id. This can be of any type. It is used to match the
+	// response with the request that it is replying to.
+	ID uint64 `json:"id"`
+}
+
+// ibcReceivePublishEvent represents the log message received from the IBC receiver contract.
+type ibcReceivePublishEvent struct {
+	ChannelID string
+	Msg       *common.MessagePublication
+}
+
+// Run is the runnable for monitoring the IBC contract on wormchain.
+func (w *Watcher) Run(ctx context.Context) error {
+	w.logger = supervisor.Logger(ctx)
+
+	errC := make(chan error)
+	defer close(errC)
+
+	// Query the contract for the chain ID to IBC connection ID mapping.
+	connectionIdMap, err := w.queryConnectionIdMap()
```

Comment:
Any connection added during runtime will not be recognized. It would be best to move this to a method / function to allow performing this periodically in the future.

Reply:
Adding new chains requires a guardian release (security feature requested by @evan-gray) so this query is only necessary at start up.

Reply:
My main thought is that there might be something whitelisted in the software that is not yet whitelisted on-chain but I'm ok with making the sacrifice of requiring a restart.